common/src/main/java/org/apache/comet/vector/CometDictionary.java

```diff
@@ -100,17 +102,21 @@ public void close() {
   }
 
   private void initialize() {
-    switch (values.getValueVector().getMinorType()) {
+    ValueVector vector = values.getValueVector();
+    switch (vector.getMinorType()) {
       case DECIMAL:
-        // We only need to copy values for decimal type as random access
-        // to the dictionary is not efficient for decimal (it needs to copy
-        // the value to a new byte array everytime).
-        binaries = new ByteArrayWrapper[numValues];
-        for (int i = 0; i < numValues; i++) {
-          // Need copying here since we re-use byte array for decimal
-          byte[] bytes = new byte[DECIMAL_BYTE_WIDTH];
-          bytes = values.copyBinaryDecimal(i, bytes);
-          binaries[i] = new ByteArrayWrapper(bytes);
+        if (values.useDecimal128
```

Comment:
When we execute natively, I think `useDecimal128` will always be true, so this won't currently improve performance?

```
        // Switch to use Decimal128 regardless of precision, since Arrow native execution
        // doesn't support Decimal32 and Decimal64 yet.
        SQLConf.get.setConfString(CometConf.COMET_USE_DECIMAL_128.key, "true")
```

Reply:
useDecimal128 is hardcoded to true in this code in `importVector`.

```
      // Native execution should always have 'useDecimal128' set to true since it doesn't support
      // other cases.
      arrayVectors += CometVector.getVector(
        importer.importVector(arrowArray, arrowSchema, dictionaryProvider),
        true, // <-- useDecimal128
        dictionaryProvider)
```

Reply:
It is true that this is an optimization only for the integer decimal case.

Reply:
I think this is still worth it. At the moment we are likely to see a fair amount of usage with only scan enabled.

Reply:
I can probably move this logic to the reader side, so that it helps for native exec scenario as well